batch-processing/pipeline.go

```diff
@@ -0,0 +1,353 @@
+	// Upload the image to S3
+	// Send the image to the result channel
+	defer wg.Done()
+	for {
+		row, ok := <-p.channels[CONVERT]
+		if !ok {
+			break
+		}
+
+		if row.Err != nil {
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		file, err := os.Open(row.Output)
+
+		if err != nil {
+			row.Err = err
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		// Upload to S3
+		path := strings.Split(row.Output, "/")
+		key := path[len(path)-1]
+
+		err = UploadToS3WithBackoff(file, key, p.config.Aws, p.maxRetries)
+
+		if err != nil {
+			row.Err = err
+			p.channels[UPLOAD] <- row
+			continue
+		}
+
+		row.S3url = fmt.Sprintf("https://%s.s3.amazonaws.com/%s", p.config.Aws.s3bucket, key)
+
+		p.channels[UPLOAD] <- row
+	}
+}
+
+func (p *Pipeline) Write(done chan bool) {
+
+	f, err := os.Create(p.config.OutputFilepath)
+	if err != nil {
+		log.Fatalf("error creating output file: %v", err)
+	}
+	defer f.Close()
+
+	w := csv.NewWriter(f)
+	defer w.Flush()
+	w.Write(OutputHeader)
+	var failedW *csv.Writer
+
+	// Create csv for failed output if param is set
+	if p.config.FailedOutputFilepath != "" {
+		failedF, err := os.Create(p.config.FailedOutputFilepath)
+		if err != nil {
+			log.Fatalf("error creating failed output file: %v", err)
+		}
+		defer failedF.Close()
+
+		failedW = csv.NewWriter(failedF)
+		defer failedW.Flush()
+		failedW.Write(FailedOutputHeader)
+	}
+
+	for {
+		row, ok := <-p.channels[UPLOAD]
+
+		if !ok {
+			break
+		}
+		rowErr := ""
+		// only add to failed output if row has an error, but not a duplicate error or image exists error
+		if row.Err != nil {
+			if p.config.FailedOutputFilepath != "" && row.Err != ErrDuplicateURL && row.Err != ErrImageExistsInS3 {
+				failedW.Write([]string{row.Url})
+			}
+			rowErr = row.Err.Error()
+		}
+
+		w.Write([]string{row.Url, row.Input, row.Output, row.S3url, rowErr})
+	}
+	done <- true
+}
+
+func (p *Pipeline) taskPicker(task string) Task {
+	switch task {
+	case DOWNLOAD:
+		return p.Download
+	case CONVERT:
+		return p.Convert
+	case UPLOAD:
+		return p.Upload
+	}
+
+	return nil
+}
+
+func (p *Pipeline) Execute() error {
+
+	start := time.Now()
+
+	done := make(chan bool)
+	go p.Write(done)
+
+	// Start the workers
+	wgTasks := &sync.WaitGroup{}
+	for _, task := range TASKS {
+		wgTasks.Add(1)
+		go func(task string) {
+			var wg sync.WaitGroup
+			for i := 0; i < p.workers[task]; i++ {
+				wg.Add(1)
+				go p.taskPicker(task)(&wg)
+			}
+			wg.Wait()
+			close(p.channels[task])
+			wgTasks.Done()
+		}(task)
+	}
+
+	// Start the reader
+	csvReader, err := OpenCSVFile(p.config.InputFilepath)
+
+	if err != nil {
+		return err
+	}
+	go p.Read(csvReader)
+
+	// Wait for all tasks to finish
+	wgTasks.Wait()
+	elapsed := time.Since(start)
+
+	log.Printf("Finished in %s", elapsed)
+
+	<-done
+
+	return nil
+}
+
+func NewPipeline(config *Config) *Pipeline {
+	return &Pipeline{
+		config:     config,
+		maxRetries: 3,
+		urlMap:     &sync.Map{},
+		channels: map[string]chan *Out{
```

Comment:
I think this is a pretty cute structure for packing everything together, but one down-side it has is that it's hard to see how each function relates to each other - from the signature:

```go
func (p *Pipeline) Convert(wg *sync.WaitGroup) {
```

it's not super clear where `Convert` gets its data from (the `Download`->`Convert` channel), or where it writes data (the `Convert`->`Upload` channel). This comes across in how many times you have to repeat `p.channels[CONVERT]` in the body of the function - and imagine if you added a new pipeline stage which did something like normalise the size of the images, you'd need to edit every place you write `p.channels[CONVERT]` to point at a new channel.

Not necessarily a big problem, but something to think about.

Reply:
I have refactored now to include `In` and `Out` channels for each function. This was the way I implemented it in the 1st and 2nd iterations, the reason I changed: 

1. Wanted to make sure all (or most at least) follow same structure and can be easily fired off without worrying about specific params, e.g. to achieve this structure:
```
	for _, task := range TASKS {
		wgTasks.Add(1)
		go func(task string) {
			var wg sync.WaitGroup
			for i := 0; i < p.workers[task]; i++ {
				wg.Add(1)
				go p.taskPicker(task)(&wg)
			}
			wg.Wait()
			close(p.channels[task])
			wgTasks.Done()
		}(task)
	}
```
2. If there were more stages added, the structure would also work well.